WooCommerce/Classes/Authentication/RoleEligibilityUseCase.swift

```diff
@@ -0,0 +1,159 @@
+    func syncEligibilityStatusIfNeeded()
+
+    /// Checks whether the current authenticated session has the correct role to manage the store.
+    /// Any error returned from the block means that the user is not eligible.
+    ///
+    /// - Parameters:
+    ///   - storeID: The dotcom site ID of the store.
+    ///   - completion: The block to be called when the check completes, with an optional RoleEligibilityError.
+    func checkEligibility(for storeID: Int64, completion: @escaping (RoleEligibilityError?) -> Void)
+
+    /// Returns the last error information encountered by the authenticated user.
+    func lastEligibilityErrorInfo() -> EligibilityErrorInfo?
+
+    /// Removes any stored error information.
+    func reset()
+}
+
+
+/// This component checks user's eligibility to access/manage the store, and *only* saves the data when the user has insufficient role.
+/// Currently, user eligibility check is performed from:
+///     1. The store picker screen, when the user tapped "Continue".
+///     2. App launch, to check if the current user is still eligible to access the store.
+///
+final class RoleEligibilityUseCase {
+    // MARK: Properties
+
+    private let stores: StoresManager
+
+    private let defaults: UserDefaults
+
+    /// The last eligibility error info encountered by the user.
+    private var lastErrorInfo: EligibilityErrorInfo? {
+        didSet {
+            defaults.setValue(lastErrorInfo?.toDictionary(), forKey: Constants.eligibilityErrorInfoKey)
+        }
+    }
+
+    // MARK: Initialization
+
+    init(stores: StoresManager = ServiceLocator.stores, defaults: UserDefaults = UserDefaults.standard) {
+        self.stores = stores
+        self.defaults = defaults
+        self.lastErrorInfo = loadLastErrorInfo()
+    }
+}
+
+// MARK: - Protocol Implementation
+
+extension RoleEligibilityUseCase: RoleEligibilityUseCaseProtocol {
+    func syncEligibilityStatusIfNeeded() {
+        guard stores.isAuthenticated, let storeID = stores.sessionManager.defaultStoreID else {
+            return
+        }
+
+        checkEligibility(for: storeID) { error in
+            guard let error = error else {
+                // If the user is eligible, reset the stored error information.
+                self.lastErrorInfo = nil
+                return
+            }
+
+            guard case .insufficientRole(let errorInfo) = error else {
+                // For errors other than `insufficientRole` (e.g. network errors), do nothing for now.
+                // In case of network errors, it's best to depend on currently available information.
+                // i.e., if the user is previously ineligible, then assume they are still ineligible now.
+                return
+            }
+
+            // store the user information locally.
+            self.lastErrorInfo = errorInfo
+        }
+    }
+
+    func checkEligibility(for storeID: Int64, completion: @escaping (RoleEligibilityError?) -> Void) {
```

Comment:
What do you think about using Result here? Maybe `.success` will better communicate meaning of nil error?

```suggestion
    func checkEligibility(for storeID: Int64, completion: @escaping (Result<Void, RoleEligibilityError>) -> Void) {
```

Reply:
Thank you for the suggestion! 👍  I've noticed before, that in the codebase we kinda have two styles to tackle _errorable_ void results – either with an optional error parameter, or with a `Void, Error` result type (as you've suggested).

However, after receiving similar feedback [here](https://github.com/woocommerce/woocommerce-ios/pull/4493#discussion_r662583600), I guess it makes sense to use the result type instead. Will update this soon!

Reply:
Resolved in [f2e1ccc](https://github.com/woocommerce/woocommerce-ios/pull/4492/commits/f2e1cccf92c216a97dcf4ab30806c9024fd2330e)! Sorry it took a while. 🙂